src/bin/pg_autoctl/keeper.c

```diff
@@ -2064,30 +2062,83 @@ keeper_refresh_other_nodes(Keeper *keeper, bool forceCacheInvalidation)
 		}
 	}
 
+	/*
+	 * In case of success, copy the current nodes array to the keeper's cache.
+	 */
+	bool success =
+		keeper_call_refresh_hooks(keeper, &newNodesArray, forceCacheInvalidation);
+
+	if (success)
+	{
+		keeper->otherNodes = newNodesArray;
+	}
+
+	return success;
+}
+
+
+/*
+ * keeper_call_refresh_hooks loops over the KeeperNodesArrayRefreshArray and
+ * calls each hook in turn. It returns true when all the hooks have returned
+ * true.
+ */
+bool
+keeper_call_refresh_hooks(Keeper *keeper,
+						  NodeAddressArray *newNodesArray,
+						  bool forceCacheInvalidation)
+{
+	bool success = true;
+
+	for (int index = 0; KeeperRefreshHooks[index]; index++)
+	{
+		KeeperNodesArrayRefreshFunction hookFun = KeeperRefreshHooks[index];
+
+		bool ret = (*hookFun)(keeper, newNodesArray, forceCacheInvalidation);
```

Comment:
I think my only concern here is that what if the first hook succeeds and has some side-effects, and the second hook fails.

We'd report this as a failure to the upper levels of the function but have done the side-effects.

Do you see any problems with that?

Reply:
I've done it that way on purpose, as usual the hook should expect to be called again when it reports false. I think it's a good way to approach these hooks, because we need to decide if we should go with the new array of nodes, and we should certainly call the hooks again with the same diff/changes when they have failed previously to install the changes...